In [22]:
import pandas as pd
import numpy as np

In [121]:
res_muni = pd.read_csv('\\Users\luisr\Desktop\Repository\IBM_courses\Personal Work\Data Frames\Work Copies\Source\saep\CSV\\saep_res_municipio.csv', sep=';')

In [122]:
res_muni.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60608 entries, 0 to 60607
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   ID_SAEB                60608 non-null  int64 
 1   ID_REGIAO              60608 non-null  int64 
 2   SIGLA_UF               60608 non-null  object
 3   ID_UF                  60608 non-null  int64 
 4   NOME_MUNICIPIO         60608 non-null  object
 5   ID_MUNICIPIO           60608 non-null  int64 
 6   ID_TIPO_REDE           60608 non-null  int64 
 7   ID_LOCALIZACAO         60608 non-null  int64 
 8   ID_CAPITAL             60608 non-null  int64 
 9   ID_SERIE               60608 non-null  int64 
 10  NU_MATRICULADOS_CENSO  60608 non-null  int64 
 11  NU_PRESENTES           60608 non-null  int64 
 12  TAXA_PARTICIPACAO      60608 non-null  object
 13  ID_DIVULGACAO          60608 non-null  int64 
 14  MEDIA_LP               60608 non-null  object
 15  MEDIA_MT           

In [123]:
res_muni.head()

ID_SAEB  ID_REGIAO SIGLA_UF  ID_UF NOME_MUNICIPIO  ID_MUNICIPIO  \
0     2011          1       AC     12     ACRELANDIA       1200013   
1     2011          1       AC     12     ACRELANDIA       1200013   
2     2011          1       AC     12     ACRELANDIA       1200013   
3     2011          1       AC     12     ACRELANDIA       1200013   
4     2011          1       AC     12     ACRELANDIA       1200013   

   ID_TIPO_REDE  ID_LOCALIZACAO  ID_CAPITAL  ID_SERIE  NU_MATRICULADOS_CENSO  \
0             2               0           2         5                     23   
1             2               0           2         9                    214   
2             2               1           2         9                    159   
3             2               2           2         5                     23   
4             2               2           2         9                     55   

   NU_PRESENTES TAXA_PARTICIPACAO  ID_DIVULGACAO MEDIA_LP MEDIA_MT  \
0            13       56,52173913              1   171,22    184,8   
1           174       81,30841121              1    225,3   240,09   
2           134       84,27672956              1   222,02    235,7   
3            13       56,52173913              1   171,22    184,8   
4            40       72,72727273              1    234,8   252,77   

  ERRO_PADRAO_LP ERRO_PADRAO_MT  
0              0              0  
1           0,71            1,5  
2              0              0  
3              0              0  
4           3,26           5,14

# Missing Values

In [124]:
cols = res_muni.columns
obj_cols = [cols[-6]]+list(cols[-4:])

obj_cols

['TAXA_PARTICIPACAO',
 'MEDIA_LP',
 'MEDIA_MT',
 'ERRO_PADRAO_LP',
 'ERRO_PADRAO_MT']

In [125]:
(res_muni[obj_cols]==' ').sum()

TAXA_PARTICIPACAO      0
MEDIA_LP             972
MEDIA_MT             972
ERRO_PADRAO_LP       972
ERRO_PADRAO_MT       972
dtype: int64

### Checking wether the 972 missing values for each column are aligned in the same 972 rows.

##### What we are going to do is compare the lines where values for one column is empty against the lines where the values in the other column is empty. If the lines are the same, it means the empity values are in the same lines. We are going to do this for each combination of columns with missing values we have. The same could be done comparing the rows where column values are not empty. They should be equal if empty values are all in the same rows.

In [126]:
for i in range(1, 5):
    for j in range(i,5):
        
        print(i, j, (res_muni[(res_muni[obj_cols[i]] == ' ')] != res_muni[(res_muni[obj_cols[j]] == ' ')]).sum().sum())

1 1 0
1 2 0
1 3 0
1 4 0
2 2 0
2 3 0
2 4 0
3 3 0
3 4 0
4 4 0


#### Conclusion: We have zero different lines for each two columns comparison so all our data missing are in the same 972 rows. So we can just drop these rows.

### Droping Missing Values

In [127]:
res_muni = res_muni[res_muni[obj_cols[1]]!=' ']

In [128]:
res_muni.shape

(59636, 18)

In [129]:
(res_muni[obj_cols]==' ').sum()

TAXA_PARTICIPACAO    0
MEDIA_LP             0
MEDIA_MT             0
ERRO_PADRAO_LP       0
ERRO_PADRAO_MT       0
dtype: int64

# Type Conversion

In [130]:
for col in obj_cols:
    res_muni[col] = res_muni[col].str.replace(',', '.')
    

In [131]:
res_muni[obj_cols] = res_muni[obj_cols].astype(float)

In [132]:
res_muni.dtypes

ID_SAEB                    int64
ID_REGIAO                  int64
SIGLA_UF                  object
ID_UF                      int64
NOME_MUNICIPIO            object
ID_MUNICIPIO               int64
ID_TIPO_REDE               int64
ID_LOCALIZACAO             int64
ID_CAPITAL                 int64
ID_SERIE                   int64
NU_MATRICULADOS_CENSO      int64
NU_PRESENTES               int64
TAXA_PARTICIPACAO        float64
ID_DIVULGACAO              int64
MEDIA_LP                 float64
MEDIA_MT                 float64
ERRO_PADRAO_LP           float64
ERRO_PADRAO_MT           float64
dtype: object

# Saving Cleaned Data

In [133]:
res_muni.to_csv('\\Users\luisr\Desktop\Repository\IBM_courses\Personal Work\Data Frames\Work Copies\Work\\res_municipio_cleaned.csv')